Mathlib/GroupTheory/Subsemigroup/Basic.lean

```diff
@@ -0,0 +1,504 @@
+* `Subsemigroup.closure` :  semigroup closure of a set, i.e.,
+  the least subsemigroup that includes the set.
+* `Subsemigroup.gi` : `closure : Set M → Subsemigroup M` and coercion `coe : Subsemigroup M → Set M`
+  form a `GaloisInsertion`;
+
+## Implementation notes
+
+Subsemigroup inclusion is denoted `≤` rather than `⊆`, although `∈` is defined as
+membership of a subsemigroup's underlying set.
+
+Note that `Subsemigroup M` does not actually require `Semigroup M`,
+instead requiring only the weaker `Mul M`.
+
+This file is designed to have very few dependencies. In particular, it should not use natural
+numbers.
+
+## Tags
+subsemigroup, subsemigroups
+-/
+
+
+-- Only needed for notation
+-- Only needed for notation
+variable {M : Type _} {N : Type _} {A : Type _}
+
+section NonAssoc
+
+variable [Mul M] {s : Set M}
+
+variable [Add A] {t : Set A}
+
+/-- `MulMemClass S M` says `S` is a type of sets `s : Set M` that are closed under `(*)` -/
+class MulMemClass (S : Type _) (M : outParam <| Type _) [Mul M] [SetLike S M] where
+  /-- A substructure satisfying `MulMemClass` is closed under multiplication. -/
+  mul_mem : ∀ {s : S} {a b : M}, a ∈ s → b ∈ s → a * b ∈ s
+#align mul_mem_class MulMemClass
+
+export MulMemClass (mul_mem)
+
+/-- `add_mem_class S M` says `S` is a type of sets `s : Set M` that are closed under `(+)` -/
+class AddMemClass (S : Type _) (M : outParam <| Type _) [Add M] [SetLike S M] where
+  /-- A substructure satisfying `AddMemClass` is closed under addition. -/
+  add_mem : ∀ {s : S} {a b : M}, a ∈ s → b ∈ s → a + b ∈ s
+#align add_mem_class AddMemClass
+
+export AddMemClass (add_mem)
+
+attribute [to_additive] MulMemClass
+
+/-- A subsemigroup of a magma `M` is a subset closed under multiplication. -/
+structure Subsemigroup (M : Type _) [Mul M] where
+  /-- The carrier of a subsemigroup. -/
+  carrier : Set M
+  /-- The product of two elements of a subsemigroup belongs to the subsemigroup. -/
+  mul_mem' {a b} : a ∈ carrier → b ∈ carrier → a * b ∈ carrier
+#align subsemigroup Subsemigroup
+
+/-- An additive subsemigroup of an additive magma `M` is a subset closed under addition. -/
+structure AddSubsemigroup (M : Type _) [Add M] where
+  /-- The carrier of an additive subsemigroup. -/
+  carrier : Set M
+  /-- The sum of two elements of an additive subsemigroup belongs to the subsemigroup. -/
+  add_mem' {a b} : a ∈ carrier → b ∈ carrier → a + b ∈ carrier
+#align add_subsemigroup AddSubsemigroup
+
+attribute [to_additive AddSubsemigroup] Subsemigroup
+
+namespace Subsemigroup
+
+@[to_additive]
+instance : SetLike (Subsemigroup M) M :=
+  ⟨Subsemigroup.carrier, fun p q h => by cases p; cases q; congr ⟩
+
+@[to_additive]
+instance : MulMemClass (Subsemigroup M) M where mul_mem := fun {_ _ _} => Subsemigroup.mul_mem' _
+
+/-- See Note [custom simps projection] -/
+@[to_additive " See Note [custom simps projection]"]
+def Simps.coe (S : Subsemigroup M) : Set M :=
+  S
+#align subsemigroup.simps.coe Subsemigroup.Simps.coe
```

Comment:
I'm afraid that you need to manually add `#align` statements for the to_additive versions...